### ISSUE TITLE
Add rule S6321: Administration services access should be restricted to specific IP addresses (APPSEC-168)

### ISSUE THREAD

**Issue by gaetan-ferry-sonarsource** [SonarSource/SonarJS#3442]
# What

The rule [S6321](https://sonarsource.github.io/rspec/#/rspec/S6321/javascript) detects AWS resources created with unfiltered network access to administration services.

This has been previously implemented for Python. See [SONARPY-1135](https://sonarsource.atlassian.net/browse/SONARPY-1135).

# Detection
### Notes

The `Connections` type is the one on which the checks are relevant.

A lot of `Construct` types have a `connections` attribute of type `Connections`. We want to detect things there too.

`ec2` is used as an alias for aws-cdk-lib/aws_ec2

### Definitions

Admin port: 22 or 3389

Bad protocol: ec2.Protocol.ALL or ec2.Protocol.TCP

Bad port:

* ec2.Port.allTcp()
* ec2.Port.allTraffic()
* ec2.Port.tcp(X) and X is an admin port
* ec2.Port.tcpRange(startPort=X, endPort=Y) with [X:Y] contains an admin port
* ec2.Port(protocol=P, fromPort=X, toPort=Y) with P is a bad protocol and [X:Y] contains an admin port

Bad peer:

* ec2.Peer.anyIpv4()
* ec2.Peer.anyIpv6()
* ec2.Peer.ipv4("0.0.0.0/0")
* ec2.Peer.ipv6("::/0")

Construct types that have a connections attribute:

* aws-cdk-lib/aws_docdb.DatabaseCluster
* aws-cdk-lib/aws_lambdaPythonAlpha.PythonFunction
* aws-cdk-lib/aws_batchAlpha.ComputeEnvironment
* aws-cdk-lib/aws_efs.FileSystem
* aws-cdk-lib/aws_lambdaGoAlpha.GoFunction
* aws-cdk-lib/aws_ecs.ExternalService
* aws-cdk-lib/aws_ecs.FargateService
* aws-cdk-lib/aws_ecs.Cluster
* aws-cdk-lib/aws_ecs.Ec2Service
* aws-cdk-lib/aws_elasticsearch.Domain
* aws-cdk-lib/aws_neptuneAlpha.DatabaseCluster
* aws-cdk-lib/aws_eks.FargateCluster
* aws-cdk-lib/aws_eks.Cluster
* aws-cdk-lib/aws_codebuild.PipelineProject
* aws-cdk-lib/aws_codebuild.Project
* aws-cdk-lib/aws_rds.DatabaseInstance
* aws-cdk-lib/aws_rds.DatabaseInstanceReadReplica
* aws-cdk-lib/aws_rds.DatabaseCluster
* aws-cdk-lib/aws_rds.ServerlessClusterFromSnapshot
* aws-cdk-lib/aws_rds.DatabaseProxy
* aws-cdk-lib/aws_rds.DatabaseInstanceFromSnapshot
* aws-cdk-lib/aws_rds.ServerlessCluster
* aws-cdk-lib/aws_rds.DatabaseClusterFromSnapshot
* aws-cdk-lib/aws_lambdaNodejs.NodejsFunction
* aws-cdk-lib/aws_fsx.LustreFileSystem
* aws-cdk-lib/aws_ec2.BastionHostLinux
* aws-cdk-lib/aws_ec2.ClientVpnEndpoint
* aws-cdk-lib/aws_ec2.Instance
* aws-cdk-lib/aws_ec2.LaunchTemplate
* aws-cdk-lib/aws_ec2.SecurityGroup
* aws-cdk-lib/aws_kinesisfirehoseAlpha.DeliveryStream
* aws-cdk-lib/aws_stepfunctionsTasks.SageMakerCreateTrainingJob
* aws-cdk-lib/aws_stepfunctionsTasks.SageMakerCreateModel
* aws-cdk-lib/aws_stepfunctionsTasks.EcsRunTask
* aws-cdk-lib/aws_redshiftAlpha.Cluster
* aws-cdk-lib/aws_opensearchservice.Domain
* aws-cdk-lib/aws_secretsmanager.HostedRotation
* aws-cdk-lib/aws_mskAlpha.Cluster
* aws-cdk-lib/triggers.TriggerFunction
* aws-cdk-lib/aws_autoscaling.AutoScalingGroup
* aws-cdk-lib/aws_syntheticsAlpha.Canary
* aws-cdk-lib/aws_cloudfront.experimental.EdgeFunction
* aws-cdk-lib/aws_lambda.Function
* aws-cdk-lib/aws_lambda.DockerImageFunction
* aws-cdk-lib/aws_lambda.SingletonFunction
* aws-cdk-lib/aws_lambda.Alias
* aws-cdk-lib/aws_lambda.Version

### Detection

For all construct types that have a connections attribute (e.g. `ec2.Instance`, or `ec2.BastionHostLinux`):

* `allowFrom` method is called on the `connections` property of a construct instance with the `other` parameter set to a bad peer and the `port` parameter set to a bad port.
* `allowFromAnyIpv4` method is called on the `connections` property of a construct instance with the `port` parameter set to a bad port

**Example**

```javascript
import { aws_ec2 as ec2 } from 'aws-cdk-lib'

const instance = new ec2.Instance(this, "default-own-security-group",{
    instanceType: nanoT2,
    machineImage: ec2.MachineImage.latestAmazonLinux(),
    vpc: vpc,
    instanceName: "test-instance"
})
instance.applyRemovalPolicy(cdk.RemovalPolicy.DESTROY)

// Code example for construct-only connections and IPv4
instance.connections.allowFrom(
    ec2.Peer.anyIpv4(), // Noncompliant
    ec2.Port.tcp(22),
    /*description*/ "Allows SSH from all IPv4"
)
instance.connections.allowFrom(
    ec2.Peer.ipv4("0.0.0.0/0"), // Noncompliant
    ec2.Port.tcpRange(/* startPort */ 1, /*endPort*/ 1024),
    /* description */ "Allows SSH and others from all IPv4"
)
instance.connections.allowFromAnyIpv4( // Noncompliant
    ec2.Port.tcp(3389),
    /* description */ "Allows Terminal Server from all IPv4"
)

instance.connections.allowFromAnyIpv4(  // Compliant
    ec2.Port.tcp(1234),
    /* description */ "Allows 1234 from all IPv4"
)

instance.connections.allowFrom(
    ec2.Peer.ipv4("192.0.2.0/24"), // Compliant
    ec2.Port.tcp(22),
    /* description */ "Allows SSH from all IPv4"
)
```

For `ec2.Connections` type:

* `allowFrom` method is called on a `Connections` instance with the `other` parameter set to a bad peer and the `port` parameter set to a bad port. 
* `allowFromAnyIpv4` method is called on a `Connections` instance with the `port` parameter set to a bad port. 
* `allowDefaultPortFrom` method is called on a `Connections` instance with the `other` parameter set to a bad peer and the `defaultPort` property of the `Connections` instance is known to be a bad port.
* `allowDefaultPortFromAnyIpv4` method is called on a `Connections` instance and the `defaultPort` property of the `Connections` instance is known to be a bad port

**Example**
```javascript
import { aws_ec2 as ec2 } from 'aws-cdk-lib'

const connection = new ec2.Connections({
    defaultPort: ec2.Port.tcp(22),
    securityGroups: [securityGroup]
})

connection.allowDefaultPortFromAnyIpv4("Allows SSH from all IPv4") // Noncompliant
connection.allowDefaultPortFrom(ec2.Peer.anyIpv4(), "Allows SSH from all IPv4") // Noncompliant

const connection2 = new ec2.Connections({
    defaultPort:ec2.Port.tcp(1234),
    securityGroups:[securityGroup2]
})

connection2.allowDefaultPortFromAnyIpv4("Allows 1234 from all IPv4") // Compliant
connection2.allowDefaultPortFrom(ec2.Peer.anyIpv4(), "Allows 1234 from all IPv4") // Compliant
```

For `ec2.SecurityGroup`:

* `addIngressRule` method is called on a `SecurityGroup` instance with the `peer` parameter set to a bad peer and the `connection` argument set to a bad port.

**Example**
``` javascript
import { aws_ec2 as ec2 } from 'aws-cdk-lib'

const securityGroup3 = new ec2.SecurityGroup(this, "custom-security-group", {
    vpc: vpc
})

securityGroup3.addIngressRule(
    ec2.Peer.anyIpv4(), // Noncompliant
    ec2.Port.tcpRange(1, 1024)
)

securityGroup3.addIngressRule(
    ec2.Peer.anyIpv4(),// Compliant, port is safe
    ec2.Port.tcpRange(1024, 1048) 
)
```

### Notes

ec2 is used as an alias for aws-cdk-lib/aws_ec2.

### Definitions
Bad protocol: "tcp", "6"

Admin port: 22 or 3389

Unsafe connection spec

* an ec2.CfnSecurityGroup.IngressProperty with ipProtocol set to "-1" and cidrIp set to "0.0.0.0/0"
* an ec2.CfnSecurityGroup.IngressProperty with ipProtocol set to "-1" and cidrIpv6 set to "::/0"
* an ec2.CfnSecurityGroup.IngressProperty with ipProtocol set to a bad protocol, cidrIp set to "0.0.0.0/0", fromPort set to X, and toPort set to Y, with [X:Y] contains an admin port
* an ec2.CfnSecurityGroup.IngressProperty with ipProtocol set to a bad protocol, cidrIpv6 set to"::/0", fromPort set to X, and toPort set to Y, with [X:Y] contains an admin port

### Detection
For `ec2.CfnSecurityGroup`:

* `ec2.CfnSecurityGroup` constructor is called with the `vpcId` attribute set, and the `securityGroupIngress` attribute is a list that contains an unsafe connection spec

**Example**
```javascript
import { aws_ec2 as ec2 } from 'aws-cdk-lib'

new ec2.CfnSecurityGroup(
    this,
    "cfn-based-security-group", {
      groupDescription: "cfn based security group",
      groupName: "cfn-based-security-group",
      vpcId: vpc.vpcId,
      securityGroupIngress: [
          {
              ipProtocol: "6",
              cidrIp: "0.0.0.0/0", // Noncompliant
              fromPort: 22,
              toPort: 22
          },
          {
              ipProtocol: "tcp",
              cidrIp: "0.0.0.0/0", // Noncompliant
              fromPort: 3389,
              toPort: 3389
          },
          {
              ipProtocol: "-1",
              cidrIpv6: "::/0" // Noncompliant
          },
          {
              ipProtocol: "6",
              cidrIp: "192.0.2.0/24", // Compliant
              fromPort: 22,
              toPort: 22
          },
          {
              ipProtocol: "tcp",
              cidrIp: "0.0.0.0/0", // Compliant, port is safe
              fromPort: 1024, 
              toPort: 1048
          }
      ]
  }
)
```

For `ec2.CfnSecurityGroupIngress`:

* `ec2.CfnSecurityGroup` constructor is called with the `ipProtocol` attribute set to a bad protocol, `cidrIp` set to “0.0.0.0/0",  `fromPort` set to X and `toPort` set to Y, with [X:Y] contains an admin port
* `ec2.CfnSecurityGroup` constructor is called with the `ipProtocol` attribute set to a bad protocol, `cidrIpv6` set to “::/0",  `fromPort` set to X and `toPort` set to Y, with [X:Y] contains an admin port
* `ec2.CfnSecurityGroup` constructor is called with the `ipProtocol` attribute set to "-1" and `cidrIp` set to “0.0.0.0/0"
* `ec2.CfnSecurityGroup` constructor is called with the `ipProtocol` attribute set to "-1" and `cidrIpv6` set to “::/0"

**Example**
```javascript
new ec2.CfnSecurityGroupIngress( 
    this,
    "ingress-all-ip-tcp-ssh", {
        ipProtocol: "tcp",
        cidrIp: "0.0.0.0/0", // Noncompliant
        fromPort: 22,
        toPort: 22,
        groupId: securityGroup.attrGroupId
})

new ec2.CfnSecurityGroupIngress(
    this,
    "ingress-all-ipv6-all-tcp", {
        ipProtocol: "-1",
        cidrIpv6: "::/0", // Noncompliant
        groupId: securityGroup.attrGroupId
})

new ec2.CfnSecurityGroupIngress(
    this,
    "ingress-all-ipv4-tcp-http", {
        ipProtocol: "6",
        cidrIp: "0.0.0.0/0",
        fromPort: 80, // Compliant
        toPort: 80,
        groupId: securityGroup.attrGroupId
})
```

# RSPEC

This [RSPEC PR](https://github.com/SonarSource/rspec/pull/1306) contains indications regarding the highlighting and messages.

**Comment by pierre-loup-tristant-sonarsource** [comment]
LGTM
